// @angular/platform-server/esm5/src/platform_state.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __decorate, __metadata, __param } from "tslib";
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable } from '@angular/core';
import { serializeDocument } from './domino_adapter';
/**
 * Representation of the current platform state.
 *
 * @publicApi
 */
import * as ɵngcc0 from '@angular/core';
var PlatformState = /** @class */ (function () {
    function PlatformState(_doc) {
        this._doc = _doc;
    }
    /**
     * Renders the current state of the platform to string.
     */
    PlatformState.prototype.renderToString = function () {
        return serializeDocument(this._doc);
    };
    /**
     * Returns the current DOM state.
     */
    PlatformState.prototype.getDocument = function () {
        return this._doc;
    };
    PlatformState = __decorate([ __param(0, Inject(DOCUMENT)),
        __metadata("design:paramtypes", [Object])
    ], PlatformState);
PlatformState.ɵfac = function PlatformState_Factory(t) { return new (t || PlatformState)(ɵngcc0.ɵɵinject(DOCUMENT)); };
PlatformState.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: PlatformState, factory: function (t) { return PlatformState.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PlatformState, [{
        type: Injectable
    }], function () { return [{ type: Object, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }]; }, null); })();
    return PlatformState;
}());
export { PlatformState };
